pagopa/io-app
# chore: [IOPID-894] Fast login opt-in BL

## Short description
This PR adds the fast login opt-in business logic

## List of changes proposed in this pull request
- [.env.local, .env.production, ts/config.ts]: added a Feature Flag for the optin. Until we implement the opt-in screen, it will be necessary to bypass the opt-in feature for fast login to work properly.
- [ 
        ts/features/common/store/reducers/index.ts,
        ts/features/fastLogin/saga/pendingActionsSaga.ts,
        ts/features/fastLogin/saga/tokenRefreshSaga.ts, 
        ts/features/fastLogin/saga/utils/index.ts,
        ts/features/fastLogin/screens/FastLoginModals.tsx, 
        ts/features/idpay/configuration/xstate/actions.ts,
        ts/features/idpay/onboarding/xstate/actions.ts,
        ts/features/idpay/payment/xstate/actions.ts,
        ts/features/idpay/unsubscription/xstate/actions.ts,
        ts/sagas/startup.ts,
        ts/store/actions/types.ts, 
        ts/store/reducers/authentication.ts,
        ts/features/idpay/onboarding/xstate/__tests__/actions.test.ts,
        ts/sagas/__tests__/initializeApplicationSaga.test.ts,
        ts/features/idpay/configuration/xstate/__tests__/actions.test.ts
]: updated import.
- ts/features/nativeLogin/store/selectors/index.ts: updated function.
- ts/features/fastLogin/saga/optInSaga.ts: a saga that will be used to manage the opt-in screen.
- ts/features/fastLogin/store/actions/index.ts: index refactored to handle opt-in and tokenRefresh actions.
- ts/features/fastLogin/store/actions/optInActions.ts: opt-in actions.
- ts/features/fastLogin/store/actions/tokenRefreshActions.ts: token refresh actions moved to this file.
- ts/features/fastLogin/store/reducers/index.ts: index refactored to handle opt-in and tokenRefresh reducers.
- ts/features/fastLogin/store/reducers/optInReducer.ts:  opt-in reducer.
- ts/features/fastLogin/store/reducers/tokenRefreshReducer.ts:  token refresh reducer moved to this file.
- ts/features/fastLogin/store/selectors/__tests__/fastLogin.test.ts: updated tests.
- ts/features/fastLogin/store/selectors/index.ts: added new selectors for opt-in and fast login status.
- ts/store/reducers/index.ts: updated to manage the opt-in status in case of logout.
- ts/store/reducers/__tests__/featureFlagWithMinAppVersionStatus.test.ts: added tests for the new function.
- ts/store/reducers/backendStatus.ts: moved `isPropertyWithMinAppVersionEnabled` to another file
- ts/store/reducers/featureFlagWithMinAppVersionStatus.ts:  Refactored `isPropertyWithMinAppVersionEnabled` to also check for any nested configurations that act as feature flag.

## How to test
Enable `FAST_LOGIN_OPTIN` and `FAST_LOGIN` in env file and enable Fast login and its opt-in config on dev server.

- Perform a login and check that the app is not using fast login features.
- Dispatch the `setFastLoginOptIn` action with a `enabled: true` payload inside the startup saga, before the authentication saga. The app should use fast login features.
- Repeat the operation with false in the payload and the app should not use fast login.
- Check that with `FAST_LOGIN_OPTIN` disabled (remote or local flag), the use of the fast login features depends only on the Fast Login FF


## Desired Behavior
```mermaid
graph TD
A[App IO Open] -->|fastLoginEnabled = true| B[fastLoginEnabled]
B -->|optInEnabled = true| C[optInEnabled]
B -->|optInEnabled = false| D[Fast Login Only]
C --> E[Show Complimentary Screen]
E -->|User Chooses Fast Login| F[Fast Login]
E -->|User Chooses Standard Login| G[Standard Login]
A -->|fastLoginEnabled = false| H[Standard Login Only]
```

Thanks @shadowsheep1 for the graph
